ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,95 @@
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "HardSoftLogicMixer.hpp"
+
+#include <stdint.h> // INT_MAX
+#include <vector>
+
+#include "multipliers.h" // instantiate_simple_soft_multiplier
+#include "odin_error.h"  // error_message
+
+HardSoftLogicMixer::HardSoftLogicMixer(const config_t& config) {
+    _allOptsDisabled = true;
+    // By default, disables all optimizations
+    for (int i = 0; i < operation_list_END; i++) {
+        this->_opts[i] = NULL;
+    }
+    parse_opt_parameters(config);
+}
+
+void HardSoftLogicMixer::parse_opt_parameters(const config_t config) {
+    if (config.mults_ratio != -1.0) {
+        this->_opts[MULTIPLY] = new MultsOpt(config.mults_ratio);
+        _allOptsDisabled = false;
+    } else if (config.exact_mults != -1) {
+        this->_opts[MULTIPLY] = new MultsOpt(config.mults_ratio);
+        _allOptsDisabled = false;
+    }
+}
+
+void HardSoftLogicMixer::note_candidate_node(nnode_t* opNode) {
+    _nodes_by_opt[opNode->type].push_back(opNode);
+}
+
+bool HardSoftLogicMixer::hardenable(nnode_t* node) {
+    return this->_opts[node->type]->hardenable(node);
+}
+
+bool HardSoftLogicMixer::enabled(nnode_t* node) {
+    return this->_opts[node->type]->enabled(node);
+}
+
+void HardSoftLogicMixer::map_deferred_blocks(netlist_t* netlist) {
+    if (_allOptsDisabled) {
+        return;
+    }
+    grid_statistics();
+    perform_optimizations(netlist);
+}
+
+void HardSoftLogicMixer::grid_statistics() {
+    if (true == _allOptsDisabled) {
+        return;
+    }
+
+    for (int i = 0; i < operation_list_END; i++) {
+        MixingOpt* opt = this->_opts[i];
+        if (opt != NULL) {
+            if (opt->enabled(NULL)) {
+                operation_list kind = opt->get_kind();
+                int blocks_needed = this->hard_blocks_needed(kind);
+                opt->set_blocks_needed(blocks_needed);
+            }
+        }
+    }
+}
+
+int HardSoftLogicMixer::hard_blocks_needed(operation_list opt) {
+    return _nodes_by_opt[opt].size();
+}
+
+void HardSoftLogicMixer::perform_optimizations(netlist_t* netlist) {
+    _opts[MULTIPLY]->assign_weights(netlist, _nodes_by_opt[MULTIPLY]);
```

Comment:
add `_opts[MULTIPLY]->set_blocks_needed(this->_nodes_by_opt[MULTIPLY].size());`

above

Reply:
I plan to create it right in the parameters

Reply:
Each mixing optimization determines the weights. we don't set it here?